src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -151,65 +127,52 @@ public FileStorage(final Path path, final ExecutorService exec) {
                     keys.size(), prefix.string(), this.dir, path, keys
                 );
                 return keys;
-            },
-            this.exec
-        );
+            }).to(SingleInterop.get()).toCompletableFuture();
     }
 
     @Override
     public CompletableFuture<Void> save(final Key key, final Content content) {
-        return CompletableFuture.supplyAsync(
+        return Single.fromCallable(
             () -> {
-                final Path tmp = Paths.get(
-                    this.dir.toString(),
-                    String.format("%s.%s.tmp", key.string(), UUID.randomUUID())
-                );
-                tmp.getParent().toFile().mkdirs();
-                return tmp;
-            },
-            this.exec
-        ).thenCompose(
-            tmp -> new File(tmp).write(
-                new OneTimePublisher<>(content),
-                this.exec,
-                StandardOpenOption.WRITE,
-                StandardOpenOption.CREATE,
-                StandardOpenOption.TRUNCATE_EXISTING
-            ).thenCompose(
-                nothing -> this.move(tmp, this.path(key))
-            ).handleAsync(
-                (nothing, throwable) -> {
-                    tmp.toFile().delete();
-                    final CompletableFuture<Void> result = new CompletableFuture<>();
-                    if (throwable == null) {
-                        result.complete(null);
-                    } else {
-                        result.completeExceptionally(throwable);
-                    }
-                    return result;
-                },
-                this.exec
-            ).thenCompose(Function.identity())
-        );
+                final Path file = this.path(key);
+                Files.createDirectories(file.getParent());
+                return file;
+            })
+            .flatMapCompletable(
+                file -> new RxFile(file, this.fls).save(Flowable.fromPublisher(content))
+            ).onErrorResumeNext(
+                throwable -> new RxFile(this.path(key), this.fls)
+                    .delete()
+                    .andThen(Completable.error(throwable))
+            )
+            .to(CompletableInterop.await())
+            .<Void>thenApply(o -> null)
+            .toCompletableFuture();
```

Comment:
@Sammers21 `save` operation becomes non-atomic after these changes. If there was some value saved for the `key` before the operation it will be lost on error. Also the file could be written partially if thread/process is terminated

Reply:
@olegmoz, there is no test covering it.

Reply:
@olegmoz, and why we would need it here?

Reply:
@Sammers21 please read comment above, we need `save` operation to be atomic. This property is used in many adapters. That's bad we lack tests for that, it should be addressed as well

Reply:
@olegmoz, if this property is "used in many adapters", why there are no tests for it?